src/evm-block-extractor/src/storage_clients/gcp_big_query.rs

```diff
@@ -0,0 +1,260 @@
+use ethers_core::types::{Block, Transaction, TransactionReceipt, H256};
+use gcp_bigquery_client::model::query_request::QueryRequest;
+use gcp_bigquery_client::model::table_data_insert_all_request::TableDataInsertAllRequest;
+use gcp_bigquery_client::model::table_data_insert_all_request_rows::TableDataInsertAllRequestRows;
+use gcp_bigquery_client::Client;
+use serde::Serialize;
+
+use super::BlockChainDB;
+use crate::constants::{BLOCKS_TABLE_ID, PROJECT_ID, RECEIPTS_TABLE_ID};
+
+#[derive(Clone)]
+/// A client for BigQuery that can be used to query and insert data
+pub struct BigQueryBlockChain {
+    client: Client,
+    /// The project ID of the BigQuery project
+    project_id: String,
+    /// The dataset ID of the BigQuery table
+    ///
+    /// Can be mainnet/testnet
+    dataset_id: String,
+}
+
+impl BigQueryBlockChain {
+    /// Creates a new BigQuery client
+    /// The service account key should be stored in the
+    /// `GCP_BLOCK_EXTRACTOR_SA_KEY` environment variable
+    pub async fn new(dataset_id: String, sa_key: String) -> anyhow::Result<Self> {
+        let service_account = yup_oauth2::parse_service_account_key(sa_key)?;
+
+        let client = Client::from_service_account_key(service_account, false).await?;
+
+        Ok(Self {
+            client,
+            project_id: PROJECT_ID.to_string(),
+            dataset_id: dataset_id.to_string(),
+        })
+    }
+
+    async fn execute_query<T: for<'de> serde::Deserialize<'de>>(
+        &self,
+        query: String,
+    ) -> anyhow::Result<T> {
+        let mut response = self
+            .client
+            .job()
+            .query(self.project_id.as_str(), QueryRequest::new(query))
+            .await?;
+
+        if response.next_row() {
+            let result_str = response
+                .get_string(0)?
+                .ok_or(anyhow::anyhow!("Expected result not found in the response"))?;
+            let result: T = serde_json::from_str(&result_str)?;
+            Ok(result)
+        } else {
+            Err(anyhow::anyhow!("No data found for the query"))
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockChainDB for BigQueryBlockChain {
+    async fn get_block_by_number(&self, block_number: u64) -> anyhow::Result<Block<Transaction>> {
+        let query = format!(
+            "SELECT body FROM `{project_id}.{dataset_id}.{table_id}` WHERE id = {block_number}",
```

Comment:
Out of curiosity, is there any protection against SQL injection in this gcp client?

Reply:
BigQuery parameterized queries are  immune to SQL injection by [design](https://cloud.google.com/bigquery/docs/parameterized-queries).

Reply:
Well... that was my point, these queries are not parameterized :)

Reply:
But this is client, when the request is sent over to gcp as a query parameter, it is immune to SQL injection.

Reply:
You are not using query parameters because all query data is hardcoded in the query string. You should use placeholders to avoid SQL injection (see: https://github.com/lquerel/gcp-bigquery-client/blob/bd82fd16cdac389507975743478840103747e6db/examples/pagination.rs#L13 )